File: app/src/main/java/net/gtr/framework/fragment/RxBaseDialogFragment.java
package net.gtr.framework.fragment;

import android.app.Activity;
import android.os.Bundle;
import android.support.annotation.IdRes;
import android.support.annotation.Nullable;
import android.support.v4.app.DialogFragment;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentTransaction;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import net.gtr.framework.activity.RxAppCompatActivity;
import net.gtr.framework.rx.Actions4SimpleDlg;
import net.gtr.framework.rx.DialogObserverHolder;
import net.gtr.framework.rx.ObserverResourceManager;

import org.reactivestreams.Subscription;

import butterknife.ButterKnife;
import io.reactivex.disposables.Disposable;

/**
 *
 * @author caroline
 * @date 2018/4/24
 */

public abstract class RxBaseDialogFragment extends DialogFragment implements DialogObserverHolder {
    protected RxAppCompatActivity mActivity;
    /**
     * use to manage resource
     */
    ObserverResourceManager observerResourceManager = new ObserverResourceManager();

    protected RxAppCompatActivity getBaseActivity() {
        return mActivity;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    public final View findViewById(@IdRes int id) {
        if (getView() == null) {
            return null;
        }
        return getView().findViewById(id);
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        if (container != null) {
            ButterKnife.bind(this, container);
        }
        return container;
    }

    @SuppressWarnings("deprecation")
    @Override
    public void onAttach(Activity activity) {
        super.onAttach(activity);
        this.mActivity = (RxAppCompatActivity) activity;
    }


    public void addFragment(int layout, Fragment fragment) {
        FragmentManager manager = getChildFragmentManager();
        FragmentTransaction transaction = manager.beginTransaction();
        transaction.replace(layout, fragment).commitAllowingStateLoss();
    }

    public void addFragment(int layout, Fragment fragment, boolean isSave) {
        FragmentManager manager = getChildFragmentManager();
        FragmentTransaction transaction = manager.beginTransaction();
        transaction.replace(layout, fragment);
        if (isSave) {
            transaction.addToBackStack(null);
        }
        transaction.commitAllowingStateLoss();
    }

    public void addSupportFragment(int layout, Fragment fragment) {
        FragmentManager manager = mActivity.getSupportFragmentManager();
        FragmentTransaction transaction = manager.beginTransaction();
        transaction.replace(layout, fragment).commitAllowingStateLoss();
    }

    @Override
    public void onDestroy() {
        clearWorkOnDestroy();
        super.onDestroy();
    }

    /**
     * {@link ObserverResourceManager#clearWorkOnDestroy()}
     */
    @Override
    public void clearWorkOnDestroy() {
        observerResourceManager.clearWorkOnDestroy();
    }

    /**
     * {@link ObserverResourceManager#addDisposable(Disposable)}
     */
    @Override
    public void addDisposable(Disposable disposable) {
        observerResourceManager.addDisposable(disposable);
    }

    /**
     * {@link ObserverResourceManager#addSubscription(Subscription)}
     */
    @Override
    public void addSubscription(Subscription subscription) {
        observerResourceManager.addSubscription(subscription);
    }

    /**
     * {@link ObserverResourceManager#removeDisposable(Disposable)}
     */
    @Override
    public void removeDisposable(Disposable disposable) {
        observerResourceManager.removeDisposable(disposable);
    }

    /**
     * {@link ObserverResourceManager#removeSubscription(Subscription)}
     */
    @Override
    public void removeSubscription(Subscription subscription) {
        observerResourceManager.removeSubscription(subscription);
    }

    @Override
    public FragmentManager getSupportFragmentManager() {
        return super.getFragmentManager();
    }

}

File: app/src/main/java/net/gtr/framework/fragment/RxBaseDialogFragment.kt
package net.gtr.framework.fragment

import android.app.Activity
import android.content.Context
import android.os.Bundle
import android.support.annotation.IdRes
import android.support.v4.app.DialogFragment
import android.support.v4.app.Fragment
import android.support.v4.app.FragmentManager
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import butterknife.ButterKnife
import io.reactivex.disposables.Disposable
import net.gtr.framework.activity.RxAppCompatActivity
import net.gtr.framework.rx.DialogObserverHolder
import net.gtr.framework.rx.ObserverResourceManager
import org.reactivestreams.Subscription

/**
 *
 * @author caroline
 * @date 2018/4/24
 */

abstract class RxBaseDialogFragment : DialogFragment(), DialogObserverHolder {
    protected var baseActivity: RxAppCompatActivity? = null
    /**
     * use to manage resource
     */
    internal var observerResourceManager = ObserverResourceManager()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
    }

    fun findViewById(@IdRes id: Int): View? {
        return if (view == null) {
            null
        } else view!!.findViewById(id)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        if (container != null) {
            ButterKnife.bind(this, container)
        }
        return container
    }

    override fun onAttach(activity: Activity?) {
        super.onAttach(activity)
        this.baseActivity = activity as RxAppCompatActivity
    }

    override fun getContext(): Context {
        return baseActivity?.getContext()!!
    }

    fun addFragment(layout: Int, fragment: Fragment) {
        val manager = childFragmentManager
        val transaction = manager.beginTransaction()
        transaction.replace(layout, fragment).commitAllowingStateLoss()
    }

    fun addFragment(layout: Int, fragment: Fragment, isSave: Boolean) {
        val manager = childFragmentManager
        val transaction = manager.beginTransaction()
        transaction.replace(layout, fragment)
        if (isSave) {
            transaction.addToBackStack(null)
        }
        transaction.commitAllowingStateLoss()
    }

    fun addSupportFragment(layout: Int, fragment: Fragment) {
        val manager = baseActivity?.supportFragmentManager
        val transaction = manager!!.beginTransaction()
        transaction.replace(layout, fragment).commitAllowingStateLoss()
    }

    override fun onDestroy() {
        clearWorkOnDestroy()
        super.onDestroy()
    }

    /**
     * [ObserverResourceManager.clearWorkOnDestroy]
     */
    override fun clearWorkOnDestroy() {
        observerResourceManager.clearWorkOnDestroy()
    }

    /**
     * [ObserverResourceManager.addDisposable]
     */
    override fun addDisposable(disposable: Disposable?) {
        observerResourceManager.addDisposable(disposable)
    }

    /**
     * [ObserverResourceManager.addSubscription]
     */
    override fun addSubscription(subscription: Subscription?) {
        observerResourceManager.addSubscription(subscription)
    }

    /**
     * [ObserverResourceManager.removeDisposable]
     */
    override fun removeDisposable(disposable: Disposable?) {
        observerResourceManager.removeDisposable(disposable)
    }

    /**
     * [ObserverResourceManager.removeSubscription]
     */
    override fun removeSubscription(subscription: Subscription?) {
        observerResourceManager.removeSubscription(subscription)
    }

    override fun getSupportFragmentManager(): FragmentManager? {
        return super.getFragmentManager()
    }

}
